conda-store-server/conda_store_server/utils.py

```diff
@@ -50,11 +50,48 @@ def chdir(directory: pathlib.Path):
         os.chdir(current_directory)
 
 
+def du(path):
+    """
+    Pure Python equivalent of du -sb
+
+    Based on https://stackoverflow.com/a/55648984/161801
+    """
+    if os.path.islink(path):
+        return os.lstat(path).st_size
+    if os.path.isfile(path):
+        st = os.lstat(path)
+        return st.st_size
+    apparent_total_bytes = 0
+    have = set()
+    for dirpath, dirnames, filenames in os.walk(path):
+        apparent_total_bytes += os.lstat(dirpath).st_size
+        for f in filenames:
+            fp = os.path.join(dirpath, f)
+            if os.path.islink(fp):
+                apparent_total_bytes += os.lstat(fp).st_size
+                continue
+            st = os.lstat(fp)
+            if st.st_ino in have:
+                continue
+            have.add(st.st_ino)
+            apparent_total_bytes += st.st_size
+        for d in dirnames:
+            dp = os.path.join(dirpath, d)
+            if os.path.islink(dp):
+                apparent_total_bytes += os.lstat(dp).st_size
+
+    # Round up
+    n_blocks = (apparent_total_bytes + 511) // 512
+    return n_blocks
```

Comment:
Will do more tests tomorrow. So far: tested on Linux against `du -sb`, returns wrong results for `/bin`. For `/bin/ls` it seems to work. I'll write a test.

Reply:
OK, this is a bit tricky.

First, we incorrectly calculate file size on macOS with `du`. There's no built-in set of options that can give us the equivalent of `-sb` AFAICT. We could install `du` from GNU coreutils, which is what is used on Linux, then we'd get the same size as our script. _But I'd suggest we just use our Python code to calculate the size on all systems._ However, it would still not match what Finder is showing, but I think it's close enough to simply ignore. Different tools calculate this differently and it's not hugely important to users unless we're in the same ballpark.

I also had to remove the round up calculation from the original script. Also, simplified it.

I also haven't tested this script with symlink/filesystem loops. Is this a safe assumption to make that we won't run into this where this is supposed to be used?

My updated code + some tests: https://gist.github.com/nkaretnikov/1a66b90a74fa805f1022e90252e54c87
Note: uncomment the TemporaryDirectory line again and remove two lines after it, which were added for testing on Windows:
```
    # with tempfile.TemporaryDirectory() as dir:
        dir = "c:\\tmpdir"
        os.makedirs(dir)
```

On Windows, I've also attempted to test against [du from SysInternals](https://learn.microsoft.com/en-us/sysinternals/downloads/du), but it also shows different info. It hasn't been updated in a while, so I just ignored that.

In general, I suggest we rely on what OS built-in "File info" tools return to debug this.

On Linux, the updated script matches the du command we're using.

Windows (updated Python du matches native File info size, Sysinternals du shows different info):
<img width="488" alt="Screen Shot 2023-10-01 at 12 55 55" src="https://github.com/conda-incubator/conda-store/assets/10469885/ad678106-0845-4f5e-8b78-9933a2f8d653">

macOS (updated Python du roughly matches native File info, built-in du returns different size):
<img width="274" alt="Screen Shot 2023-10-01 at 13 12 57" src="https://github.com/conda-incubator/conda-store/assets/10469885/2723803e-cb1c-49d4-b43a-4d4e8f86ad4c">

macOS (updated Python du matches du from GNU coreutils, installed via brew):
<img width="456" alt="Screen Shot 2023-10-01 at 13 19 23" src="https://github.com/conda-incubator/conda-store/assets/10469885/107fe34d-b3b6-42be-b6e5-452563ba7c2a">

Note: creating symlinks on Windows requires having Developer Mode on:
```
Settings > Privacy & Security > For Developers and turn Developer Mode to on
```

Reply:
When I tested it on my Mac, du and this function gave the same numbers. I'm not sure if the subtle differences of blocks matter much (it matters if there are sparse or compressed files, but I doubt those would show up). The main thing is that we treat hard links correctly. 

Note that in my tests, this function is significantly slower than `du`, so we should consider whether it's worth using it over du.

Reply:
I tested it on my conda-store-state with a couple of environments:

```
$ du -sAB1 conda-store-state/
1111657	conda-store-state/
```

```
>>> from conda_store_server.utils import du
>>> du('conda-store-state')
1111657
```

That's on my Mac. I also can confirm that some of the environments do share files via hard-links:

```
$ ls -i conda-store-state/admin/b82cde5b5489ceffd8a8589ebd73f20a9f4836260b18295f49e057c441b235dc-20231005-213031-558981-2-test2/lib/libsqlite3.0.dylib
256048028 conda-store-state/admin/b82cde5b5489ceffd8a8589ebd73f20a9f4836260b18295f49e057c441b235dc-20231005-213031-558981-2-test2/lib/libsqlite3.0.dylib
$ ls -i conda-store-state/admin/99108419ad0fd922fdeff9bbc434b58d41f68e3f923a83f6a7ab19568463bc82-20231005-211948-613237-1-test/lib/libsqlite3.0.dylib
256048028 conda-store-state/admin/99108419ad0fd922fdeff9bbc434b58d41f68e3f923a83f6a7ab19568463bc82-20231005-211948-613237-1-test/lib/libsqlite3.0.dylib
```

I don't remember if I tested it on Linux, so it's possible there's a discrepancy there.

Reply:
OK, I just checked on Linux, and apparently the `du -sb` command we are using is 512 times bigger than the macOS `du -sAB1` (despite the man pages implying that they would be the same) and this `du()` function. I guess we should fix them both to account for that, since the code is already written assuming bytes instead of blocks.

Reply:
I pushed a fix for this. Note that the numbers will be off a little bit (by less than 512) on Mac because `du` rounds up to the nearest multiple of 512 when converting bytes to blocks.

However, we still need to figure out something for this `du()` function. I just tested it against my main `~/anaconda` directory and I had to exit out of it after several minutes. I think it might be accidentally quadratic. 

There is a Windows `du` command someone at https://learn.microsoft.com/en-us/sysinternals/downloads/du. Maybe we should just package that as a conda package so we can just use it directly (or maybe it already is packaged?).

Reply:
Actually even regular `du -sAB1 ~/anaconda` is slow on my machine. Maybe it's just too hard to deal with that many hard links. 

But I just noticed that the way conda-store is using this, it gathers the stats for each prefix as it is created. I need to double check it, but I think it actually isn't accounting for hard links across environments correctly anyway.

Reply:
Honestly, I'd appreciate some guidance from @costrouc or someone else on how disk_usage is actually used in conda-store before I feel comfortable with the du stuff here.

Reply:
> There is a Windows du command someone at https://learn.microsoft.com/en-us/sysinternals/downloads/du. Maybe we should just package that as a conda package so we can just use it directly (or maybe it already is packaged?).

I've already tried it. See above where I post my du test results. Sysinternals du printed wrong results for me, compared to Windows file explorer. The output also differs from Linux/macOS du.

Reply:
Oh I didn't notice you tested it already.

Regarding what the Finder returns, I would be curious to know exactly why the discrepancy exists. You can access it programmatically with AppleScript

```
osascript -e 'tell application "Finder" to get physical size of folder "Macintosh HD:Users:aaronmeurer:Documents:conda-store:conda-store-server:conda-store-state"'
6.6859008E+8
```

(I highly recommend using an LLM to help you write AppleScript)